### PR TITLE
CCM-1572: Add nhs number checkdigit validation

### PR DESCRIPTION
**Pull request by seanmarjason-nhs**
## Summary
* Routine Change
* :exclamation: Breaking Change
* :robot: Operational or Infrastructure Change
* :sparkles: New Feature
* :warning: Potential issues that might be caused by this change

Add any other relevant notes or explanations here. **Remove this line if you have nothing to add.**

## Reviews Required
* [x] Dev
* [ ] Test
* [ ] Tech Author
* [ ] Product Owner

## Review Checklist
:information_source: This section is to be filled in by the **reviewer**.

* [ ] I have reviewed the changes in this PR and they fill all or part of the acceptance criteria of the ticket, and the code is in a mergeable state.
* [ ] If there were infrastructure, operational, or build changes, I have made sure there is sufficient evidence that the changes will work.
* [ ] I have ensured the changelog has been updated by the submitter, if necessary.

### CODE REVIEW

File: proxies/shared/resources/jsc/MessageBatches.Create.Validate.js [hunk 1]
```diff
@@ -28,6 +28,28 @@ const isUndefined = (val) => {
     return typeof val === "undefined";
 }
 
+const validateNhsNumber = (nhsNumber, nhsNumberRegex) => {
+  if (!nhsNumberRegex.test(nhsNumber)) {
+    return false
```
Comment (PhilipSkinner) [inline proxies/shared/resources/jsc/MessageBatches.Create.Validate.js:33]: `;`

File: proxies/shared/resources/jsc/MessageBatches.Create.Validate.js [hunk 2]
```diff
@@ -28,6 +28,28 @@ const isUndefined = (val) => {
...
+  }
+
+  const checkDigit = nhsNumber.substring(9);
```
Comment (PhilipSkinner) [inline proxies/shared/resources/jsc/MessageBatches.Create.Validate.js:36]: `const digits = nhsNumber.split('');`
`const checkDigit = digits[9];`

?

File: proxies/shared/resources/jsc/MessageBatches.Create.Validate.js [hunk 3]
```diff
@@ -28,6 +28,28 @@ const isUndefined = (val) => {
...
+
+  const digits = []
+  for (var i = 0; i <= 8; i++) {
```
Comment (PhilipSkinner) [inline proxies/shared/resources/jsc/MessageBatches.Create.Validate.js:39]:
```
for (var i = 0; i < 9; i++) {
 digits[i] = parseInt(digits[i]);
}
```

Comment (PhilipSkinner) [reply proxies/shared/resources/jsc/MessageBatches.Create.Validate.js:39]: Also do it in a single pass:

```
var total = 0;
const digits = nhsNumber.split('');
for (var i = 0; i < 9; i++) {
  total += (parseInt(digits[i]) * (10 - i));
}

const remainder = total % 11;
return (11 - remainder) % 11 === parseInt(digits[9]);
```

File: proxies/shared/resources/jsc/MessageBatches.Create.Validate.js (hunk 2 above)
Comment (seanmarjason-nhs) [reply proxies/shared/resources/jsc/MessageBatches.Create.Validate.js:36]: Split apparently not supported with string separators (or blank separators) in RhinoJS

File: proxies/shared/resources/jsc/MessageBatches.Create.Validate.js [hunk 4]
```diff
@@ -240,6 +258,15 @@ function tooFewItemsError(pointer) {
   )
 }
 
+function invalidNhsNumberError(pointer) {
+  return createErrorObject(
+    "CM_INVALID_NHS_NUMBER",
+    "Invalid nhs number",
+    "The value provided in the nhs number property is not a valid length or does not contain valid checksum.",
```
Comment (PhilipSkinner) [inline proxies/shared/resources/jsc/MessageBatches.Create.Validate.js:265]: `nhsNumber` is the property name

File: specification/communications-manager.yaml [hunk 5]
```diff
@@ -255,6 +255,7 @@ paths:
             | `CM_MISSING_VALUE` | Missing property | The property at the specified location is required, but was not present in the request. |
             | `CM_NULL_VALUE` | Property cannot be null | The property at the specified location is required, but a null value was passed in the request. |
             | `CM_INVALID_VALUE` | Invalid value | The property at the specified location does not allow this value. |
+            | `CM_INVALID_NHS_NUMBER` | Invalid nhs number | The value provided in the nhs number property is not a valid length or does not contain valid [checksum](https://www.datadictionary.nhs.uk/attributes/nhs_number.html). |
```
Comment (PhilipSkinner) [inline specification/communications-manager.yaml:258]: This URL should be in the `links` section returned by the error, plus we need to update the documentation on the `nhsNumber` field to include a link to this page aswell. I'd update this string to be the same as that output by the system above (post update).

File: specification/schemas/UnableToProcessResponse.yaml [hunk 6]
```diff
@@ -50,7 +50,6 @@ properties:
                 - /data/attributes/messageBatchReference
                 - /data/attributes/routingPlanId
                 - /data/attributes/messages
-                - /data/attributes/messages/0/recipient/nhsNumber
```
Comment (PhilipSkinner) [inline specification/schemas/UnableToProcessResponse.yaml:53]: This should be reverted.

**Review by PhilipSkinner — CHANGES_REQUESTED** [review]
I think the documentation and language could do with some work, the change to behaviour all looks good.

File: proxies/shared/resources/jsc/MessageBatches.Create.Validate.js [hunk 7]
```diff
@@ -236,10 +256,21 @@ function tooFewItemsError(pointer) {
...
   )
 }
 
+function invalidNhsNumberError(pointer) {
+  return createErrorObject(
+    "CM_INVALID_NHS_NUMBER",
+    "Invalid nhs number",
+    "The value provided in this nhsNumber field is not a valid NHS number.",
+    pointer,
+    {"nhsNumbers": "https://www.datadictionary.nhs.uk/attributes/nhs_number.html"}
```
Comment (PhilipSkinner) [inline proxies/shared/resources/jsc/MessageBatches.Create.Validate.js:270]: nice :)

**Review by PhilipSkinner — APPROVED** [review]

**Review by Kennedy1265 — APPROVED** [review]

**Pull request merged**